doc/CFG.md

```diff
@@ -0,0 +1,108 @@
+
+
+# Sonar-PHP Control Flow Graph
+
+## Definition
+
+A CFG is a graph of [basic blocks](https://en.wikipedia.org/wiki/Basic_block) (from now on, we will use the term 'block'), where block is sequence of statements executed linearly without branching.
+
+### Blocks
+
+#### Block attributes
+
+* successors - set of blocks that are executed after specific block
+* predecessors - set of blocks that are executed before specific block
+* syntactic successor
+  - an imaginary successor which exists only for blocks ending with unconditional jumps (`break`, `continue`, `return`, `goto`, `throw`)
+  - it is the "normal" successor in case the jump would be omitted
+  - for non-jump blocks, it is `null`
+* elements
+  - list of statements (represented as AST node) inside the block which are executed sequentially
+  - `return` or jump like `break`, `continue` will be the last statement in the elements list
+
+#### Block types
+
+Different implementations of blocks:
+
+* Simple block - one or multiple successors
+* Branching block
+  - has 2 successors - one TRUE successor and one FALSE successor
+  - contains the branching tree - e.g. IfStatementTree or loop trees
```

Comment:
I don't understand this item. Using `IfStatementTree` and loop trees seems inconsistent to me.

Reply:
Loops and If are generating branching block because they split the control flow to the true and false branches